python/cudf_polars/cudf_polars/experimental/rapidsmpf/nodes.py

```diff
@@ -0,0 +1,436 @@
+from cudf_polars.experimental.rapidsmpf.dispatch import generate_ir_sub_network
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncIterator
+
+    from rapidsmpf.streaming.core.channel import Channel
+    from rapidsmpf.streaming.core.context import Context
+
+    from cudf_polars.experimental.rapidsmpf.dispatch import SubNetGenerator
+
+
+@asynccontextmanager
+async def shutdown_on_error(
+    ctx: Context, *channels: Channel[Any]
+) -> AsyncIterator[None]:
+    """
+    Shutdown on error for rapidsmpf.
+
+    Parameters
+    ----------
+    ctx
+        The context.
+    channels
+        The channels to shutdown.
+    """
+    # TODO: This probably belongs in rapidsmpf.
+    try:
+        yield
+    except Exception:
+        await asyncio.gather(*(ch.shutdown(ctx) for ch in channels))
+        raise
+
+
+@define_py_node()
+async def default_node_single(
+    ctx: Context,
+    ir: IR,
+    ch_out: ChannelPair,
+    ch_in: ChannelPair,
+) -> None:
+    """
+    Single-channel default node for rapidsmpf.
+
+    Parameters
+    ----------
+    ctx
+        The context.
+    ir
+        The IR node.
+    ch_out
+        The output ChannelPair.
+    ch_in
+        The input ChannelPair.
+
+    Notes
+    -----
+    Chunks are processed in the order they are received.
+    """
+    async with shutdown_on_error(ctx, ch_in.data, ch_out.data):
+        while (msg := await ch_in.data.recv(ctx)) is not None:
+            chunk = TableChunk.from_message(msg)
+            seq_num = chunk.sequence_number
+            df = await asyncio.to_thread(
+                ir.do_evaluate,
+                *ir._non_child_args,
+                DataFrame.from_table(
+                    chunk.table_view(),
+                    list(ir.children[0].schema.keys()),
+                    list(ir.children[0].schema.values()),
+                    chunk.stream,
+                ),
+            )
+            chunk = TableChunk.from_pylibcudf_table(
+                seq_num, df.table, chunk.stream, exclusive_view=True
+            )
+            await ch_out.data.send(ctx, Message(chunk))
+
+        await ch_out.data.drain(ctx)
+
+
+@define_py_node()
+async def default_node_multi(
+    ctx: Context,
+    ir: IR,
+    ch_out: ChannelPair,
+    chs_in: tuple[ChannelPair, ...],
+    bcast_indices: list[int],
+) -> None:
+    """
+    Pointwise node for rapidsmpf.
+
+    Parameters
+    ----------
+    ctx
+        The context.
+    ir
+        The IR node.
+    ch_out
+        The output ChannelPair (metadata already sent).
+    chs_in
+        Tuple of input ChannelPairs (metadata already received).
+    bcast_indices
+        The indices of the broadcasted children.
+
+    Notes
+    -----
+    Input chunks are aligned for evaluation.
+    """
+    # TODO: Use multiple streams
+    async with shutdown_on_error(ctx, *[ch.data for ch in chs_in], ch_out.data):
+        seq_num = 0
+        n_children = len(chs_in)
+        accepting_data = True
```

Comment:
Could you walk me through why we need this `accepting_data` flag?

Reply:
One specific question: IIUC we have data coming in on multiple channels, and we need to make sure we have the i-th chunk / table for each channel before we can run `ir.do_evaluate` on that list of tables. So we stage these chunks in `staged_chunks`.

Can / should we clear those from `staged_chunks` immediately after the `ir.do_evaluate` so that we aren't keeping those tables alive and using up memory unnecessarily?